src/ReflectionFileNamespace.php

```diff
@@ -83,6 +83,9 @@ class ReflectionFileNamespace
      */
     public function __construct($fileName, $namespaceName, Namespace_ $namespaceNode = null)
     {
+        if (!is_string($fileName)) {
```

Comment:
Ok... I'm happy to show you the three places that call this constructor, but the fact that you're asking indicates a failure to communicate on my part. I apologize for being unclear.

Here are the 3 places that call this constructor:


[.../parser-reflection/src/ReflectionFile.php](https://github.com/loren-osborn/parser-reflection/blob/295b2da02320cf978694172cd4ddc9a993cedfd8/src/ReflectionFile.php#L155):
```php
  153                  $namespaceName = $topLevelNode->name ? $topLevelNode->name->toString() : '';
  154  
  155:                 $namespaces[$namespaceName] = new ReflectionFileNamespace(
  156                      $this->fileName,
  157                      $namespaceName,
```

[.../parser-reflection/src/ValueResolver/NodeExpressionResolver.php](https://github.com/loren-osborn/parser-reflection/blob/295b2da02320cf978694172cd4ddc9a993cedfd8/src/ValueResolver/NodeExpressionResolver.php#L233):
```php
  231                  $fileName      = $this->context->getFileName();
  232                  $namespaceName = $this->resolveScalarMagicConstNamespace();
  233:                 $fileNamespace = new ReflectionFileNamespace($fileName, $namespaceName);
  234                  if ($fileNamespace->hasConstant($constantName)) {
  235                      $constantValue = $fileNamespace->getConstant($constantName);
```

[...](https://github.com/loren-osborn/parser-reflection/blob/295b2da02320cf978694172cd4ddc9a993cedfd8/src/ValueResolver/NodeExpressionResolver.php#L502)
```php
  500              $namespaceName = $this->resolveScalarMagicConstNamespace();
  501  
  502:             $fileNamespace = new ReflectionFileNamespace($fileName, $namespaceName);
  503              return $fileNamespace->getClass($className);
  504          }
```

That said, I think this largely misses the point. Let me see if I can clear this up a little bit:

- **First, an apology:** When you first brought these two changes up as issues, I think I took it a bit personally, and doubled down by adding unit tests for my type checking code. I was hoping that after trying to explain myself, why I made these changes would be clear, and I tried to defer to @lisachenko as an arbitrator. This basically blew up in my face, and I failed to address your concerns about these changes directly. I'm sorry. _Mea culpa._

- **Relevance:** The only reason I included these checks in this PR was because Scrutinizer flagged them as issues. While I try not to take anything that Scrutinizer says as gospel, in this case it has a point: These are public API entry points, and can be called with invalid arguments. Since we can't add PHP 7 type hints here without breaking PHP 5 compatibility, I felt it was appropriate to add these type checks to fix Scrutinizer's issue.

- **Performance:** You have repeatedly brought up the concern of this check impacting performance, but in these two instances it just isn't applicable. Of course I agree that any operation takes some amount overhead. The problem is that these are checks that only occur on the construction of infrequently created objects that **often** result in loading and parsing a file. Even if the parsed file cache is 80% effective, and we're pulling data from a solid state drive, or even the OS cache, checking the data type of a single function parameter has to be **at least** 3 to 5 orders of magnitude smaller reading and parsing a file 20% of the time. I agree that it contributes to constructor complexity which impacts maintainability, and both of those need to be kept in check, but this has **no appreciable** impact on performance.

- **Addressing this from where the method is called:** I think this is where our biggest misunderstanding lies.
    + My goal is to eliminate the issues being reported by Scrutinizer.
    + Scrutinizer is reporting two almost identical issues on two constructors with identically defined `$fileName` parameters.
    + One of those constructors is never called by this project's code, while the other is called three times.
    + **If** ensuring that all calling code was properly type hinted, then we should expect that the constructor that is never called (so all zero of its callers are properly type-hinted) would not trigger any Scrutinizer issue.
    + Since the uncalled constructor **does** trigger a Scrutinizer issue, I think it's safe to assume that ensuring the callers are properly type hinted (which I believe they already are) will have any impact on Scrutinizer reporting this issue.
    + **Also,** as this is a public API, I see no harm in a bit of sanity checking of parameters.

- **Summary:** The decision is ultimately yours, so if you ask me to get rid of these two checks, and the unit tests to support them, I will, but I think that asking that really misses the point of this PR.

**So, your choices are:**

1. I can leave these two `if (is_string())` checks, along with the corresponding unit tests, in place.
2. Or I can remove them.